lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,399 @@
+	}
+
+	return &childListener{
+		Listener:            listener,
+		childNetworkingFile: listenerFD,
+	}, nil
+}
+
+// ListenX11 requests a local XServer listener from the networking subprocess.
+func (p *Process) ListenX11(ctx context.Context, req X11Request) (net.Listener, error) {
+	listenerFD, err := p.sendRequest(ctx, Request{
+		Operation:  NetworkingOperationListenX11,
+		X11Request: req,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(listenerFD.file)
+	if err != nil {
+		listenerFD.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	return &childListener{
+		Listener:            listener,
+		childNetworkingFile: listenerFD,
+	}, nil
+}
+
+// childNetworkingFile is a networking file corresponding to a networking
+// resource opened in the child process, with a lifetime associated with
+// the per-request connection.
+type childNetworkingFile struct {
+	conn *net.UnixConn
+	file *os.File
+}
+
+func (f *childNetworkingFile) Close() error {
+	return trace.NewAggregate(f.conn.Close(), f.file.Close())
+}
+
+// childConn is a net.Conn backed by a connection in the child process.
+type childConn struct {
+	childNetworkingResource *childNetworkingFile
+	net.Conn
+}
+
+// Close the local file connection and the child networking resources behind it.
+func (c *childConn) Close() error {
+	return trace.NewAggregate(c.Conn.Close(), c.childNetworkingResource.Close())
+}
+
+// childListener is a net.Listener backed by a listener in the child process.
+type childListener struct {
+	childNetworkingFile *childNetworkingFile
+	net.Listener
+}
+
+// Close the local file listener and the child networking resources behind it.
+func (l *childListener) Close() error {
+	return trace.NewAggregate(l.Listener.Close(), l.childNetworkingFile.Close())
+}
+
+// sendRequest sends a networking request to the networking process and waits
+// for a file corresponding to an open network connection or listener.
+func (p *Process) sendRequest(ctx context.Context, req Request) (*childNetworkingFile, error) {
+	slog.DebugContext(ctx, "Sending networking request to child process", "request", req)
+
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// We can use a stream instead of a datagram because we only expect bytes
+	// or a file descriptor in response, not both.
+	requestConn, remoteConn, err := uds.NewSocketpair(uds.SocketTypeStream)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	remoteFD, err := remoteConn.File()
+	remoteConn.Close()
+	if err != nil {
+		requestConn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	if _, _, err = uds.WriteWithFDs(p.conn, jsonReq, []*os.File{remoteFD}); err != nil {
+		requestConn.Close()
+		remoteFD.Close()
+		return nil, trace.Wrap(err)
+	}
+	remoteFD.Close()
+
+	// Read in another goroutine so we can cancel it if the networking process stops or the context closes.
```

Comment:
Why? Just tie the context to the closing of the connection with a `context.AfterFunc` (which returns a function you can call to undo the tying).

Reply:
We don't want to close the per-request connection if the request succeeds, as the networking resources's lifetime is tied to the connection. We could add a conditional close if that's better? Seems iffy.

```go
func (p *Process) sendRequest(ctx context.Context, req Request) (f *childNetworkingFile, err error) {
    context.AfterFunc(func(){
        if err != nil {
            reqConn.Close()
        )
    })
```

Reply:
You can cancel a context AfterFunc, which we should take advantage of.